File: adapters/python/avro_adapters/item/rpc_client.py
"""RPC adapter: implements Repository over HTTP+JSON.

Delegates all calls to a remote server, making local↔remote swap transparent
to the application core and service layer. The wire protocol mirrors the
messages in ports/item/repository.avpr.
"""

import json
import urllib.error
import urllib.request

from .model import Item
from .repository import NotFound


class RpcClient:
    """Adapter that satisfies Repository by calling a remote ItemRepository server."""

    def __init__(self, addr: str) -> None:
        self._base = f"http://{addr}"

    def _post(self, path: str, body: dict) -> dict | list | None:
        data = json.dumps(body).encode()
        req = urllib.request.Request(
            self._base + path,
            data=data,
            headers={"Content-Type": "application/json"},
        )
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            if e.code == 404:
                payload = json.loads(e.read())
                raise NotFound(payload.get("id", "?"))
            raise

    def get(self, id: str) -> Item:
        data = self._post("/item/get", {"id": id})
        return Item(**data)

    def list(self) -> list[Item]:
        data = self._post("/item/list", {})
        return [Item(**d) for d in (data or [])]

    def save(self, item: Item) -> None:
        self._post("/item/save", {"id": item.id, "name": item.name})

    def delete(self, id: str) -> None:
        self._post("/item/delete", {"id": id})

File: adapters/go/item/rpc_client.go
package item

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

// RpcClient is an adapter that implements Repository over HTTP+JSON.
// It delegates all calls to a remote server, making local↔remote swap
// transparent to the application core and service layer.
//
// The wire protocol mirrors the messages in ports/item/repository.avpr.
type RpcClient struct {
	base   string
	client *http.Client
}

func NewRpcClient(addr string) *RpcClient {
	return &RpcClient{base: "http://" + addr, client: &http.Client{}}
}

func (c *RpcClient) post(path string, body, out any) error {
	b, err := json.Marshal(body)
	if err != nil {
		return err
	}
	resp, err := c.client.Post(c.base+path, "application/json", bytes.NewReader(b))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		var nf NotFoundError
		json.NewDecoder(resp.Body).Decode(&nf)
		return &nf
	}
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("server error: %s", resp.Status)
	}
	if out != nil {
		return json.NewDecoder(resp.Body).Decode(out)
	}
	return nil
}

func (c *RpcClient) Get(id string) (*Item, error) {
	var result Item
	err := c.post("/item/get", map[string]string{"id": id}, &result)
	return &result, err
}

func (c *RpcClient) List() ([]*Item, error) {
	var result []*Item
	err := c.post("/item/list", struct{}{}, &result)
	return result, err
}

func (c *RpcClient) Save(it *Item) error {
	return c.post("/item/save", it, nil)
}

func (c *RpcClient) Delete(id string) error {
	return c.post("/item/delete", map[string]string{"id": id}, nil)
}
